### PR TITLE
[extension/ackextension] Create Ack extension interface

### PR DESCRIPTION
**Pull request by zpzhuSplunk**
**Description:** <Describe what has changed.>
Adding interface and other related files to serve as placeholder for the Ack extension proposed in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26376. Implementation will be added after this PR gets merged. 

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26376

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: extension/ackextension/factory.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	"go.opentelemetry.io/collector/extension"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/bearertokenauthextension/internal/metadata"
+)
+
+const (
+	defaultStorageType = "in-memory"
+)
+
+func NewFactory() extension.Factory {
```
Comment (splunkericl) [inline extension/ackextension/factory.go:16]: add test for the factory

File: extension/ackextension/internal/metadata/generated_status.go
Comment (splunkericl) [inline extension/ackextension/internal/metadata/generated_status.go:1]: Shouldn't this file to be automatically generated by `mdatagen`? Or was this manually added?

File: .chloggen/ack_extension_interface.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'new_component'
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: 'ackextension'
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Adding interface and other related files to serve as placeholder for the Ack extension proposed in #26376. Implementation will be added after this PR gets merged."
```
Comment (splunkericl) [inline .chloggen/ack_extension_interface.yaml:10]: This note will be exposed in release notes. We should re-phrase this for end users to read. maybe just:
`Adding an interface for Acknowledgement extension`

**Comment by splunkericl** [comment]
Hey @atoulme , we are thinking of breaking the change into 3 parts:
1. add interface, factory and skeleton files
2. add extension implementation
3. use extension inside hec receiver

Do you think this will be the right approach to contribute this piece? Or do you prefer to combine everything in one shot?

File: extension/ackextension/config.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ackextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/ackextension"
+
+// Config defines configuration for ack extension
+type Config struct {
+	// StorageType defines the storage type of the extension. Currently planned for In memory type. Future consideration is disk type.
+	StorageType string `mapstructure:"StorageType,omitempty"`
```
Comment (dmitryax) [inline extension/ackextension/config.go:9]: Can we use `storage` option instead pointing to a storage extension similar to what we do in https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper#persistent-queue. If not provided, memory storage is used.

**Comment by dmitryax** [comment]
> we are thinking of breaking the change into 3 parts

Right, please follow this approach

File: extension/ackextension/config.go (hunk 3 above)
Comment (zpzhuSplunk) [reply extension/ackextension/config.go]: Okay. I changed the type to `*component.ID`

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -81,6 +81,7 @@ exporter/syslogexporter/                                 @open-telemetry/collect
 exporter/tencentcloudlogserviceexporter/                 @open-telemetry/collector-contrib-approvers @wgliang @yiyang5055
 exporter/zipkinexporter/                                 @open-telemetry/collector-contrib-approvers @MovieStoreGuy @astencel-sumo @crobert-1
 
+extension/ackextension/                                  @open-telemetry/collector-contrib-approvers @atoulme
```
Comment (dmitryax) [inline .github/CODEOWNERS:84]: We need more code owners from your side. This requires OTel org membership. @splunkericl I see you have made enough contributions to apply for membership. Can you submit an issue similar to https://github.com/open-telemetry/community/issues/1915?

Comment (dmitryax) [reply .github/CODEOWNERS:84]: @zpzhuSplunk @splunkericl, while the component is in the development phase, I believe you should be able to add yourselves in the code owner list even without the Opentlemetry membership

Comment (splunkericl) [reply .github/CODEOWNERS:84]: Sure we can.

File: extension/ackextension/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+## Configuration
+
+```yaml
+extensions:
+  ack:
+      storagetype: "in-memory"
+
+receivers:
+  splunk_hec:
+    ackExtension: ack
```
[the remark this answers is not in the document]
Comment (zpzhuSplunk) [reply extension/ackextension/README.md]: fixed!

**Review by dmitryax — APPROVED** [review]
LGTM. Just needs updating .github/CODEOWNERS

File: extension/ackextension/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+<!-- end autogenerated section -->
+
+This extension allows acking of data upon successful processing. The upstream agent can choose to send event again
+if ack fails. 
+## Configuration
+
+```yaml
+extensions:
+  ack:
+      storagetype: "in-memory"
```
Comment (fatsheep9146) [inline extension/ackextension/README.md:22]: small nit, I think we'd better keep the indent consistent

```
extensions:
  ack:
    storagetype: "in-memory"

receivers:
  splunk_hec:
    ack_extensions: ack
```

**Review by fatsheep9146 — APPROVED** [review]

File: extension/ackextension/README.md (hunk 6 above)
Comment (dmitryax) [reply extension/ackextension/README.md:22]: Right. Please also update it to use `storage` instead of `storagetype`

**Comment by fatsheep9146** [comment]
@zpzhuSplunk 
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31225/commits/bdc786b58137351ae63918a734e8764d10f1a85d
sorry for this commit, I just want to make a comments but made a commit instead, feel free to revert it.

File: extension/ackextension/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+<!-- end autogenerated section -->
+
+This extension allows acking of data upon successful processing. The upstream agent can choose to send event again
+if ack fails. 
+## Configuration
+
+```yaml
+extensions:
+  ack:
+    storageID:
```
Comment (dmitryax) [inline extension/ackextension/README.md:22]: Let's use `storage` to align with other components e.g. https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queue_sender.go#L40C41-L40C48
```suggestion
    storage: 
```

File: extension/ackextension/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+## Configuration
+
+```yaml
+extensions:
+  ack:
+    storageID: 
+
+receivers:
+  splunk_hec:
+    ack_extensions: ack
```
Comment (dmitryax) [inline extension/ackextension/README.md:26]:
```suggestion
    ack_extension: ack
```

**Review by dmitryax — APPROVED** [review]

**Pull request merged**